dot/parachain/prospective-parachains/fragment-chain/fragment_chain.go

```diff
@@ -0,0 +1,1196 @@
+	if setOfCandidates == nil {
+		setOfCandidates = make(map[parachaintypes.CandidateHash]any)
+	}
+	setOfCandidates[candidate.candidateHash] = struct{}{}
+	c.byOutputHead[candidate.outputHeadDataHash] = setOfCandidates
+
+	c.byCandidateHash[candidate.candidateHash] = candidate
+	return nil
+}
+
+func (c *CandidateStorage) removeCandidate(candidateHash parachaintypes.CandidateHash) {
+	entry, ok := c.byCandidateHash[candidateHash]
+	if !ok {
+		return
+	}
+
+	delete(c.byCandidateHash, candidateHash)
+
+	if setOfCandidates, ok := c.byParentHead[entry.parentHeadDataHash]; ok {
+		delete(setOfCandidates, candidateHash)
+		if len(setOfCandidates) == 0 {
+			delete(c.byParentHead, entry.parentHeadDataHash)
+		}
+	}
+
+	if setOfCandidates, ok := c.byOutputHead[entry.outputHeadDataHash]; ok {
+		delete(setOfCandidates, candidateHash)
+		if len(setOfCandidates) == 0 {
+			delete(c.byOutputHead, entry.outputHeadDataHash)
+		}
+	}
+}
+
+func (c *CandidateStorage) markBacked(candidateHash parachaintypes.CandidateHash) {
+	entry, ok := c.byCandidateHash[candidateHash]
+	if !ok {
+		fmt.Println("candidate not found while marking as backed")
+	}
+
+	entry.state = Backed
+}
+
+func (c *CandidateStorage) contains(candidateHash parachaintypes.CandidateHash) bool {
+	_, ok := c.byCandidateHash[candidateHash]
+	return ok
+}
+
+// candidates returns an iterator over references to the stored candidates, in arbitrary order.
+func (c *CandidateStorage) candidates() iter.Seq[*CandidateEntry] {
+	return func(yield func(*CandidateEntry) bool) {
+		for _, entry := range c.byCandidateHash {
+			if !yield(entry) {
+				return
+			}
+		}
+	}
+}
+
+func (c *CandidateStorage) headDataByHash(hash common.Hash) *parachaintypes.HeadData {
+	// first, search for candidates outputting this head data and extract the head data
+	// from their commitments if they exist.
+	// otherwise, search for candidates building upon this head data and extract the
+	// head data from their persisted validation data if they exist.
+
+	if setOfCandidateHashes, ok := c.byOutputHead[hash]; ok {
+		for candidateHash := range setOfCandidateHashes {
+			if candidate, ok := c.byCandidateHash[candidateHash]; ok {
+				return &candidate.candidate.Commitments.HeadData
+			}
+		}
+	}
+
+	if setOfCandidateHashes, ok := c.byParentHead[hash]; ok {
+		for candidateHash := range setOfCandidateHashes {
+			if candidate, ok := c.byCandidateHash[candidateHash]; ok {
+				return &candidate.candidate.PersistedValidationData.ParentHead
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *CandidateStorage) possibleBackedParaChildren(parentHeadHash common.Hash) iter.Seq[*CandidateEntry] {
+	return func(yield func(*CandidateEntry) bool) {
+		seqOfCandidateHashes, ok := c.byParentHead[parentHeadHash]
+		if !ok {
+			return
+		}
+
+		for candidateHash := range seqOfCandidateHashes {
+			if entry, ok := c.byCandidateHash[candidateHash]; ok && entry.state == Backed {
+				if !yield(entry) {
+					return
+				}
+			}
+		}
+	}
+}
```

Comment:
Why do we need to return an iterator? Can't we return the slice instead, simplifying the code and making it more readable?

Reply:
Here we want to give the caller to iterate over a list of candidates that is owned by Candidate Storage, the caller is not interested in owning a slice of items but just want the `possibles` to use them in another part of the code. So I think the best way to express that is using golang iterators.